Obsidian.API/Containers/AbstractContainer.cs

```diff
@@ -0,0 +1,147 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Obsidian.API
+{
+    public abstract class AbstractContainer : IEnumerable<ItemStack>, IEnumerable
+    {
+        protected ItemStack?[] items;
+
+        public int Size { get; protected set; }
+
+        public InventoryType Type { get; }
+
+        public ChatMessage? Title { get; set; }
+
+        public Guid Uuid { get; private set; } = Guid.NewGuid();
+
+        public List<IPlayer> Viewers { get; private set; } = new List<IPlayer>();
+
+        public ItemStack? this[int index] { get => this.items[index]; set => this.items[index] = value; }
+
+        public AbstractContainer(int size) : this(size, InventoryType.Custom) { }
+
+        internal AbstractContainer(int size, InventoryType type)
+        {
+            this.Size = size;
+            this.Type = type;
+
+            this.items = new ItemStack?[size];
```

Comment:
Since `items` length equals to `Size`, do we need `Size` at all? Using `items.Length` simplifies the class and avoids bound checks when used in for loops.

```cs
public int Size => items.Length;
```